File: cpp/jump_game.cc
/*
 * Filename: jump_game.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 27 Jul 2015 12:01:38 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		bool canJump(vector<int>& nums) {
			if (nums.size() < 2) {
				return true;
			}

			int stt = 0;
			int end = 0;
			
			while (end < nums.size() - 1) {
				int eend = end;

				for (vector<int>::size_type i = stt; i <= end; i++) {
					if (i + nums[i] >= nums.size() - 1) {
						return true;
					}

					eend = max(eend, i + nums[i]);
				}

				if (eend == end) {
					return false;
				}

				stt = end + 1;
				end = eend;
			}
		}

		return false;
};

File: java/JumpGame.java
/*
 * Filename: JumpGame.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 15 May 2015 08:39:00 PM EDT
 */

public class Solution {
	public boolean canJump(int[] nums) {
		if (nums == null || nums.length <= 1) {
			return true;
		}

		int i = nums.length - 1;
		int cnt = 0;
		int mark = nums.length;
		boolean rres = false;
		boolean res = true;
		boolean no0 = true;

		while (i >= 0 && nums[i] == 0) {
			cnt++;
			mark = i;
			rres = false;
			i--;
		}

		if (mark != nums.length) {
			while (i >= 0 && nums[i] != 0) {
				if (nums[i] >= mark - i + cnt - 1) {
					rres |= true;
				}
				i--;
			}
			res &= rres;
			cnt = 0;
		} else {
			while (i >= 0 && nums[i] != 0) {
				i--;
			}
		}

		while (i >= 0) {
			while (i >= 0 && nums[i] == 0) {
				cnt++;
				mark = i;
				rres = false;
				i--;
			}

			while (i >= 0 && nums[i] != 0) {
				if (nums[i] > mark - i + cnt - 1) {
					rres |= true;
				}
				i--;
			}

			res &= rres;
			cnt = 0;
		}

		return res;
	}
}
